File: apps/web-student/src/lib/assignments.test.ts
// Sprint 9 F-1 — assignment copy helper tests.
//
// Two functions, two contracts:
//  - progressBucket() drives which pill colour the row shows
//  - formatDueAt() drives the human-readable due-date copy

import { describe, expect, test } from "vitest";

import {
  formatDueAt,
  progressBucket,
  type Assignment,
} from "./assignments";

function _a(over: Partial<Assignment> = {}): Assignment {
  return {
    id: "a-1",
    cohortId: "c-1",
    tenantId: null,
    title: "Test",
    description: null,
    createdBy: "t-1",
    dueAt: null,
    publishedAt: "2026-04-28T00:00:00Z",
    createdAt: "2026-04-28T00:00:00Z",
    updatedAt: "2026-04-28T00:00:00Z",
    myCompletedAt: null,
    myCorrectCount: null,
    myTotalCount: null,
    ...over,
  };
}

describe("progressBucket", () => {
  test("completed bucket when student has completed timestamp", () => {
    const a = _a({
      myCompletedAt: "2026-04-29T10:00:00Z",
      dueAt: "2026-04-30T00:00:00Z",
    });
    expect(progressBucket(a, new Date("2026-05-15T00:00:00Z"))).toBe("completed");
  });

  test("overdue when due in past and not completed", () => {
    const a = _a({ dueAt: "2026-04-15T00:00:00Z" });
    expect(progressBucket(a, new Date("2026-04-28T00:00:00Z"))).toBe("overdue");
  });

  test("due-soon when within 24h and not completed", () => {
    const a = _a({ dueAt: "2026-04-29T05:00:00Z" });
    expect(progressBucket(a, new Date("2026-04-28T18:00:00Z"))).toBe("due-soon");
  });

  test("open when due > 24h away", () => {
    const a = _a({ dueAt: "2026-05-10T00:00:00Z" });
    expect(progressBucket(a, new Date("2026-04-28T00:00:00Z"))).toBe("open");
  });

  test("open when no due date", () => {
    const a = _a({ dueAt: null });
    expect(progressBucket(a, new Date("2026-04-28T00:00:00Z"))).toBe("open");
  });

  test("completed wins over overdue (completed late)", () => {
    const a = _a({
      dueAt: "2026-04-15T00:00:00Z",
      myCompletedAt: "2026-04-20T00:00:00Z",
    });
    expect(progressBucket(a, new Date("2026-04-28T00:00:00Z"))).toBe("completed");
  });
});

describe("formatDueAt", () => {
  test("empty string when no due date", () => {
    expect(formatDueAt(_a({ dueAt: null }))).toBe("");
  });

  test("'Due today' when due within 24h", () => {
    const a = _a({ dueAt: "2026-04-28T22:00:00Z" });
    expect(formatDueAt(a, new Date("2026-04-28T08:00:00Z"))).toBe("Due today");
  });

  test("'Due tomorrow' when due 24-48h away", () => {
    const a = _a({ dueAt: "2026-04-29T22:00:00Z" });
    expect(formatDueAt(a, new Date("2026-04-28T08:00:00Z"))).toBe("Due tomorrow");
  });

  test("'Due in Nd' for further-future due", () => {
    const a = _a({ dueAt: "2026-05-05T00:00:00Z" });
    expect(formatDueAt(a, new Date("2026-04-28T00:00:00Z"))).toBe("Due in 7d");
  });

  test("'Overdue (yesterday)' when 1 day past", () => {
    const a = _a({ dueAt: "2026-04-27T08:00:00Z" });
    expect(formatDueAt(a, new Date("2026-04-28T08:00:00Z"))).toBe(
      "Overdue (yesterday)",
    );
  });

  test("'Overdue (Nd ago)' when more than 1 day past", () => {
    const a = _a({ dueAt: "2026-04-20T00:00:00Z" });
    expect(formatDueAt(a, new Date("2026-04-28T00:00:00Z"))).toBe(
      "Overdue (8d ago)",
    );
  });

  test("'Due <date>' once student has completed (no urgency framing)", () => {
    const a = _a({
      dueAt: "2026-04-25T00:00:00Z",
      myCompletedAt: "2026-04-26T10:00:00Z",
    });
    const got = formatDueAt(a, new Date("2026-04-28T00:00:00Z"));
    expect(got).toMatch(/^Due /);
    expect(got).not.toContain("Overdue");
  });
});

File: apps/mobile/test/assignments_test.dart
// Sprint 9 F-2 — mobile assignment helper tests.

import 'package:flutter_test/flutter_test.dart';

import 'package:adaptive_learning_mobile/api/assignments.dart';

Assignment _a({
  String? dueAt,
  String? completedAt,
  int? correct,
  int? total,
}) {
  return Assignment.fromJson({
    'id': 'a-1',
    'cohortId': 'c-1',
    'tenantId': null,
    'title': 'Test',
    'description': 'Desc',
    'createdBy': 't-1',
    'dueAt': dueAt,
    'publishedAt': '2026-04-28T00:00:00Z',
    'createdAt': '2026-04-28T00:00:00Z',
    'updatedAt': '2026-04-28T00:00:00Z',
    'myCompletedAt': completedAt,
    'myCorrectCount': correct,
    'myTotalCount': total,
  });
}

void main() {
  group('progressBucket', () {
    test('completed when myCompletedAt set', () {
      final a = _a(
        dueAt: '2026-04-30T00:00:00Z',
        completedAt: '2026-04-29T00:00:00Z',
      );
      expect(progressBucket(a, DateTime.utc(2026, 5, 15)),
          ProgressBucket.completed);
    });

    test('overdue when dueAt is past and not completed', () {
      final a = _a(dueAt: '2026-04-15T00:00:00Z');
      expect(
          progressBucket(a, DateTime.utc(2026, 4, 28)), ProgressBucket.overdue);
    });

    test('dueSoon when within 24h', () {
      final a = _a(dueAt: '2026-04-29T05:00:00Z');
      expect(progressBucket(a, DateTime.utc(2026, 4, 28, 18)),
          ProgressBucket.dueSoon);
    });

    test('open when no due date', () {
      final a = _a();
      expect(progressBucket(a, DateTime.utc(2026, 4, 28)), ProgressBucket.open);
    });

    test('open when due > 24h away', () {
      final a = _a(dueAt: '2026-05-10T00:00:00Z');
      expect(progressBucket(a, DateTime.utc(2026, 4, 28)), ProgressBucket.open);
    });
  });

  group('formatDueAt', () {
    test('empty when no dueAt', () {
      expect(formatDueAt(_a(), DateTime.utc(2026, 4, 28)), '');
    });

    test('"Due today" when within 24h', () {
      final a = _a(dueAt: '2026-04-28T22:00:00Z');
      expect(formatDueAt(a, DateTime.utc(2026, 4, 28, 8)), 'Due today');
    });

    test('"Due tomorrow" when 24-48h away', () {
      final a = _a(dueAt: '2026-04-29T22:00:00Z');
      expect(formatDueAt(a, DateTime.utc(2026, 4, 28, 8)), 'Due tomorrow');
    });

    test('"Due in Nd" further out', () {
      final a = _a(dueAt: '2026-05-05T00:00:00Z');
      expect(formatDueAt(a, DateTime.utc(2026, 4, 28)), 'Due in 7d');
    });

    test('"Overdue (yesterday)" 1 day past', () {
      final a = _a(dueAt: '2026-04-27T08:00:00Z');
      expect(formatDueAt(a, DateTime.utc(2026, 4, 28, 8)),
          'Overdue (yesterday)');
    });

    test('"Overdue (Nd ago)" further past', () {
      final a = _a(dueAt: '2026-04-20T00:00:00Z');
      expect(formatDueAt(a, DateTime.utc(2026, 4, 28)), 'Overdue (8d ago)');
    });

    test('completed always uses "Due <date>" framing', () {
      final a = _a(
        dueAt: '2026-04-25T00:00:00Z',
        completedAt: '2026-04-26T10:00:00Z',
      );
      final got = formatDueAt(a, DateTime.utc(2026, 4, 28));
      expect(got, startsWith('Due '));
      expect(got.contains('Overdue'), isFalse);
    });
  });

  group('Assignment.fromJson', () {
    test('parses null optional fields', () {
      final a = _a();
      expect(a.dueAt, isNull);
      expect(a.myCompletedAt, isNull);
      expect(a.myCorrectCount, isNull);
    });

    test('parses populated optional fields', () {
      final a = _a(
        dueAt: '2026-05-01T00:00:00Z',
        completedAt: '2026-04-30T18:00:00Z',
        correct: 4,
        total: 5,
      );
      expect(a.dueAt, DateTime.utc(2026, 5, 1));
      expect(a.myCorrectCount, 4);
      expect(a.myTotalCount, 5);
    });
  });
}
